### PR TITLE
APP-281: [iOS] Create "GPIO Control" app

### PR DESCRIPTION
**Pull request by vilimets**
@abohomol please review

### CODE REVIEW

File: gpiocontrol/source/objc/GPIOControlDemo/KaaClientManager.m [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                                          delegate:delegate];
+}
+
+- (void)detachEndpoint:(NSString *)endpointId delegate:(id<OnDetachEndpointOperationDelegate>)delegate {
+    EndpointKeyHash *keyHash = [[EndpointKeyHash alloc] initWithKeyHash:endpointId];
+    [self.kaaClient detachEndpointWithKeyHash:keyHash delegate:delegate];
+}
+
+- (KAAProfileEmptyData *)getProfile
+{
```
Comment (abohomol) [inline gpiocontrol/source/objc/GPIOControlDemo/KaaClientManager.m]: Minor code style violation

**Comment by abohomol** [comment]
LGTM

File: gpiocontrol/source/objc/GPIOControlDemo/KaaClientManager.h [hunk 2]
```diff
@@ -16,22 +16,20 @@
...
+@interface KaaClientManager : NSObject
 
 @property (nonatomic, strong) volatile id<KaaClient> kaaClient;
 
-+ (id<KaaClient>)getClient;
-- (void)attachUser;
-+ (void)setUpEventDelegate:(id<RemoteControlECFDelegate>)delegate;
-+ (void)sendDeviceInfoRequestToAll;
++ (KaaClientManager *)sharedManager;
+- (void)sendDeviceInfoRequestToAll:(id <RemoteControlECFDelegate>)delegate;
```
Comment (am1kk) [inline gpiocontrol/source/objc/GPIOControlDemo/KaaClientManager.h]: I think it's better to call method like sendDeviceInfoRequestToAllWithDelegate:

File: gpiocontrol/source/objc/GPIOControlDemo/DevicesTableViewController.m [hunk 3]
```diff
@@ -40,126 +36,60 @@ @implementation DevicesTableViewController
...
 #pragma mark - Actions
 
 - (IBAction)plusButtonPressed:(UIBarButtonItem *)sender {
+    UIAlertController *alert = nil;
     if ([[[ConnectivityChecker alloc] init] isConnected]) {
-        UIAlertController *alert = [UIAlertController
-                                    alertControllerWithTitle:@"Access token"
-                                    message:@"Enter access token"
-                                    preferredStyle:UIAlertControllerStyleAlert];
+         alert = [UIAlertController alertControllerWithTitle:@"Access token"
```
Comment (am1kk) [inline gpiocontrol/source/objc/GPIOControlDemo/DevicesTableViewController.m]: Is this line has correct alignment?

File: gpiocontrol/source/objc/GPIOControlDemo/DevicesTableViewController.m [hunk 4]
```diff
@@ -172,24 +102,26 @@ - (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)
...
         Device *currentDevice = self.devices[indexPath.row];
         self.deletingIndexPath = indexPath;
-        [kaaClient detachEndpointWithKeyHash:[[EndpointKeyHash alloc] initWithKeyHash:currentDevice.kaaEndpointId] delegate:self];
+        [[KaaClientManager sharedManager] detachEndpoint:currentDevice.kaaEndpointId
+                                                delegate:self];
     }
 }
 
-#pragma mark - Delegates methods
+#pragma mark - Kaa Delegates methods
```
Comment (am1kk) [inline gpiocontrol/source/objc/GPIOControlDemo/DevicesTableViewController.m]: It's better to call mark with corresponding protocol name(e.g. RemoteControlECFDelegate, OnDetachEndpointOperationDelegate, OnAttachEndpointOperationDelegate) for every methods block.

File: gpiocontrol/source/objc/GPIOControlDemo/GPIOTableViewController.m [hunk 5]
```diff
@@ -68,29 +66,24 @@ - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(N
...
+        [self presentViewController:[ConnectionAlert noConnectionAlert]
+                           animated:YES
+                         completion:nil];
     }
 }
 
+- (void)sendToggleRequest:(UISwitch *)sender {
+    RemoteControlECFGpioStatus *status = [self.gpioStatusArray objectAtIndex:sender.tag];
+    status.status = sender.on;
+    id <KaaClient> client = [KaaClientManager sharedManager].kaaClient;
```
Comment (am1kk) [inline gpiocontrol/source/objc/GPIOControlDemo/GPIOTableViewController.m]: Please don't put a space between an object type and the protocol it conforms to

File: gpiocontrol/source/objc/GPIOControlDemo/KaaClientManager.m [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
++ (KaaClientManager *)sharedManager {
+    static KaaClientManager *sharedMyManager = nil;
+    @synchronized(self) {
+        if (sharedMyManager == nil)
+            sharedMyManager = [[self alloc] init];
+    }
+    return sharedMyManager;
+}
+
+- (id)init {
```
Comment (am1kk) [inline gpiocontrol/source/objc/GPIOControlDemo/KaaClientManager.m]: Return type should be instancetype.

File: gpiocontrol/source/objc/GPIOControlDemo/KaaClientManager.m [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+                                          delegate:delegate];
+}
+
+- (void)detachEndpoint:(NSString *)endpointId delegate:(id<OnDetachEndpointOperationDelegate>)delegate {
+    EndpointKeyHash *keyHash = [[EndpointKeyHash alloc] initWithKeyHash:endpointId];
+    [self.kaaClient detachEndpointWithKeyHash:keyHash delegate:delegate];
+}
+
+- (KAAProfileEmptyData *)getProfile
+{
```
Comment (am1kk) [inline gpiocontrol/source/objc/GPIOControlDemo/KaaClientManager.m]: Place curly braces on the same line before

File: gpiocontrol/source/objc/GPIOControlDemo/KaaClientManager.m [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    return self.profile;
+}
+
+- (void)onAttachResult:(UserAttachResponse *)response {
+    NSLog(@"User attach resultType: %u", response.result);
+}
+
+- (void)onStarted{
```
Comment (am1kk) [inline gpiocontrol/source/objc/GPIOControlDemo/KaaClientManager.m]: Place whitespace between method name and curly brace.

File: gpiocontrol/source/objc/GPIOControlDemo/KaaClientManager.m [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+- (void)onStartFailureWithException:(NSException *)exception {
+    NSLog(@"Kaa client startup failure. %@", exception);
+}
+- (void)onPaused {
+    NSLog(@"Kaa client paused");
+}
+- (void)onPauseFailureWithException:(NSException *)exception {
+    NSLog(@"Kaa client pause failure. %@", exception);
+}
+- (void)onResume{
```
Comment (am1kk) [inline gpiocontrol/source/objc/GPIOControlDemo/KaaClientManager.m]: Place whitespace between method name and curly brace.

**Comment by am1kk** [comment]
LGTM

**Pull request merged**